### ISSUE TITLE
SSD/VOC2007 validation doesn't evaluate all images when gt_difficults is not used

### ISSUE THREAD

**Issue by lgov** [dmlc/gluon-cv#536]
Hi, 

I'm finetuning a pretrained 'ssd_512_resnet50_v1_voc' network, using more or less the train_ssd.py sample code provided here:
https://gluon-cv.mxnet.io/build/examples_detection/finetune_detection.html

Using gluon-cv master @ HEAD

During validation, I find that the update function of VOC07MApMetric doesn't take into account all the samples in a batch when using multiple GPU's.

```
eval_metric = VOC07MApMetric(iou_thresh=0.5, class_names=CLASSES)
eval_metric.update(det_bboxes, det_ids, det_scores, gt_bboxes, gt_ids)
```
So gt_diffults = None in the call to update.

The update function loops over all the samples in the batch and the calculates the metric over all samples.         

```
for pred_bbox, pred_label, pred_score, gt_bbox, gt_label, gt_difficult in zip(
                *[as_numpy(x) for x in [pred_bboxes, pred_labels, pred_scores,
                                        gt_bboxes, gt_labels, gt_difficults]]):
```

However, even though as_numpy(pred_bboxes) contains 8 results (batch size = 8), it only iterates two times. 

I notice that the length of as_numpy(gt_difficults) is 2 instead of 8. As the zip() iterator has only as many elements as the shortest iterator (gt_difficults in this case), that explains why it stops after two iterations.

The root cause is this line:
```
        if gt_difficults is None:
            gt_difficults = [None for _ in gt_labels]
```

We expect gt_difficults to contain [None] * batch_size elements. 
That works when using one GPU only, but when using multiple GPU's, gt_labels has an extra dimension, so the effective result of is that gt_difficults = [None] * num_of_gpus.

**Comment by zhreshold** [comment]
closed by #537

**Comment by HaydenFaulkner** [comment]
This issue appears to go further than what has been mentioned above. When validating with the voc metric on voc data, the input into the metric's `update()` function are lists of `len() = num_gpus` which contain the `mxnet.NDArray`'s of size `(B, N, ...)`. However the `update()` function expects the input to have the first dimension as the batch size, not the number of gpus. This caused a different evaluation results for different batch sizes. If the `batch_size / num_gpus = 1` the desired results are given, otherwise if the batch_size > 1 (or in fact `gt_labels[0].shape[0] != 1`) there is incorrect overwriting of `gt_difficults` with `None`s:

For example when `n_gpu=1, batch_size=2`:
```
if isinstance(gt_labels, list):  # will be True
    if len(gt_difficults) != len(gt_labels) * gt_labels[0].shape[0]:  # will be True --> 1 != 1 * 2
         gt_difficults = [None] * len(gt_labels) * gt_labels[0].shape[0]  # now gt_difficults = [None, None]
```
Two ways to fix this could be:
1. To ensure that the input to the `update()` function meets the function requirements and is of shape `(B, N, ...)` which can be easily done by applying the `as_numpy()` function seen in the voc metric to all inputs (in the training scripts):
```
eval_metric.update(as_numpy(det_bboxes), as_numpy(det_ids), as_numpy(det_scores), as_numpy(gt_bboxes), as_numpy(gt_ids), as_numpy(gt_difficults))
```

2. Fix it in the `update()` function itself, however I'm not sure what the purpose of the offending code is, so unsure if a fix would break its purpose? Simply removing the 3 lines fix the issue:
```
if isinstance(gt_labels, list):
    if len(gt_difficults) != len(gt_labels) * gt_labels[0].shape[0]:
        gt_difficults = [None] * len(gt_labels) * gt_labels[0].shape[0]
```
Or maybe something like below would be more appropriate?
```
if isinstance(gt_labels, list):
    if len(gt_difficults) * gt_difficults[0].shape[0] != len(gt_labels) * gt_labels[0].shape[0]:
        gt_difficults = [None] * len(gt_labels) * gt_labels[0].shape[0]
```
